### PR TITLE
Editor preview

### PR DESCRIPTION
**Pull request by elkuku**
This will add an "editor" with preview functionality (as seen on GitHub :wink:)

The "lesson" to learn here might be using the markdown service in the controller, making a simple AJAX request and respond with a `JsonResponse` object from the controller.

Please bear with me since I am very new to Symfony.

<del>Currently the loading message ("Loading preview...") is hard coded inside the JavaScript file.
Does Symfony provide a better way to translate strings in JavaScript files or should I pass the string as a parameter?</del>

Hope you find it useful :wink:

### CODE REVIEW

**Comment by elkuku** [comment]
So this is actually not working as expected.
There might be some additional Symfony lessons I have to learn.
Maybe something is missing in this [empty class](https://github.com/elkuku/symfony-demo/blob/da4f281c4842fd40fb355b89fdc4f312a8a2fb89/src/AppBundle/Form/Type/EditorType.php)?

Sorry for the noise.. :(

**Comment by elkuku** [comment]
OK I found the missing piece, even the unit tests are passing now :)

This seems to be a pitfall for inexperienced Symfony developers. Not sure what may have caused the misbehavior...

File: app/Resources/views/admin/blog/new.html.twig [hunk 1]
```diff
@@ -1,5 +1,16 @@
...
 
+{% block javascripts %}
+    {{ parent() }}
+    <script>
+        $('a[data-toggle="tab"]').on('click', function (e) {
+            if ('#previewId' == $(e.target).attr('href')) {
+                SymfonyDemo.preview('#post_content', '#previewId', '{{ path('preview') }}');
+            }
+        });
+    </script>
```
Comment (yceruto) [inline app/Resources/views/admin/blog/new.html.twig]: You can create a template file to do that and avoid duplicated code in `edit.html.twig`.

File: src/AppBundle/Controller/Admin/BlogController.php [hunk 2]
```diff
@@ -189,4 +190,26 @@ public function deleteAction(Request $request, Post $post)
 
         return $this->redirectToRoute('admin_post_index');
     }
+
+    /**
+     * Converts a markdown string to HTML.
+     *
+     * @Route("/preview", name="preview")
```
Comment (yceruto) [inline src/AppBundle/Controller/Admin/BlogController.php]: You can name this route `admin_post_content_preview` to be consistent with this action and others routes?

File: src/AppBundle/Form/Type/EditorType.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+use Symfony\Component\Form\Extension\Core\Type\TextareaType;
+
+/**
+ * Defines a custom form field type used to display a markdown editor with preview functionality.
+ *
+ * See http://symfony.com/doc/current/cookbook/form/create_custom_field_type.html
+ *
+ * @author Nikolai Plath <der.el.kuku@gmail.com>
+ */
+class EditorType extends AbstractType
```
Comment (yceruto) [inline src/AppBundle/Form/Type/EditorType.php]: If only one value here is to override the block name used to render the form type then, use the `block_name` option instead (bonus plus to this PR) and remove entirely this form type.

For example:
```php
// AppBundle\Form\PostType

->add('content', null, [
    'block_name' => 'editor',
])
```
```twig
{# form/fields.html.twig #}

{% block _post_editor_widget %}
    ...
    {{ block('textarea_widget') }}
    ...
{% endblock %}
```
http://symfony.com/doc/current/reference/forms/types/form.html#block-name
http://symfony.com/doc/current/form/form_customization.html#form-customization-sidebar
http://symfony.com/doc/current/form/form_themes.html#form-template-blocks

**Comment by elkuku** [comment]
@yceruto Thanks for the review. Everything makes perfect sense. 
Thanks also for the links. Very much appreciated :wink:

File: src/AppBundle/Controller/Admin/BlogController.php [hunk 4]
```diff
@@ -189,4 +190,26 @@ public function deleteAction(Request $request, Post $post)
...
+        $response = [];
+        $text = $request->request->get('text');
+
+        if (!$text) {
+            $response['data'] = 'Nothing to preview...';
+        } else {
+            $response['data'] = $this->get('markdown')->toHtml($text);
+        }
+
+        return new JsonResponse($response);
```
Comment (elkuku) [inline src/AppBundle/Controller/Admin/BlogController.php]: Cool

File: src/AppBundle/Form/Type/EditorType.php (hunk 3 above)
Comment (elkuku) [inline src/AppBundle/Form/Type/EditorType.php]: Very cool :stuck_out_tongue_winking_eye:

File: web/js/main.js [hunk 5]
```diff
@@ -66,4 +66,25 @@
...
+         * Render a Markdown input to formatted HTML
+         *
+         * @param {String} text
+         * @param {String} preview
+         * @param {String} previewUrl
+         */
+        preview: function (text, preview, previewUrl) {
+            var out = $(preview);
+
+            out.html('Loading preview...');
```
Comment (elkuku) [inline web/js/main.js]: I added a CSS loading animation to avoid the translations..

File: src/AppBundle/Controller/Admin/BlogController.php [hunk 6]
```diff
@@ -18,6 +18,7 @@
 use Sensio\Bundle\FrameworkExtraBundle\Configuration\Security;
 use Symfony\Bundle\FrameworkBundle\Controller\Controller;
 use Symfony\Component\Form\Extension\Core\Type\SubmitType;
+use Symfony\Component\HttpFoundation\JsonResponse;
```
Comment (yceruto) [inline src/AppBundle/Controller/Admin/BlogController.php]: unused import?

Comment (voronkovich) [inline src/AppBundle/Controller/Admin/BlogController.php]: @elkuku, Why do you import this class? You don't use it in the `previewAction`.

**Comment by elkuku** [comment]
Thanks @voronkovich and @yceruto seems that your eyes are sharper than any of those automated testing tools :)

**Comment by elkuku** [comment]
Is there anything else I should do here?
BTW.. test failures are not related to the code changes here (I hope) :wink:

**Comment by elkuku** [comment]
Sorry, but I just found #69 :wink:

**Comment by noniagriconomie** [comment]
👍🏻 for this, 94noni is my personal account
And today i showed to someone how to deal with ajax

**Comment by javiereguiluz** [comment]
I'm closing this because we already have an Ajax feature (see #613) and because this would complicate the application too much. Let's remember that we are not making something comparable to WordPress. The blog application is an excuse to show Symfony code. We don't acre about the blog system and we don't care if it doesn't have some features that you find in other CMS or blog systems.

@elkuku I'm REALLY sorry for closing this without merging it. I know it's awful to dedicate time to something that it's not merged. That's why I always recommend to open an issue first to discuss about the feature instead of opening a pull request directly. I hope you understand this decision. Thanks!

**Pull request closed without merging**